### DISCUSSION TITLE
Performance

### DISCUSSION THREAD

**Discussion by sregg** [MobileNativeFoundation/discussions#147] [General]
My company started to seriously look into app performance.

We started by tracking app start time as the cold start of our app was getting quite slow.

We also instrumented a few critical screens to measure their load time (from the navigation action to the final render).

We're also looking at navigation latency for all screens.

We're tracking p50 and p90 for all these metrics.

We started to track them week by week but realized we got more insights on improvements and regressions when tracking by version. 

So far we're only tracking in prod but we're planning on being more proactive and run instrumented e2e tests on a device farm to catch issues earlier.

I'm curious to hear what kind of metrics you folks are tracking, what percentiles, what reference (time or version) and when (pre or post release).

Cheers!

**Comment by mattrobmattrob** [comment]
Handful of useful discussions to reference from the past:

- https://github.com/MobileNativeFoundation/discussions/discussions/37
- https://github.com/MobileNativeFoundation/discussions/discussions/146
- https://github.com/MobileNativeFoundation/discussions/discussions/133

**Comment by fbaptista** [comment]
Hello ⚡ , this topic passionates me, so here are my first two cents. 
At Apptim we help companies measuring different KPIs on CI in pre-release stage. 

While metrics highly depends on the industry, the most common ones we are asked are:
* Startup time (cold, warm, hot). In android specially the [ReportFullyDrawn()](https://developer.android.com/reference/android/app/Activity#reportFullyDrawn()) event
* App size (and also weight per modules and method count)
* Memory (games take this very seriously specially when supporting low-end devices)
* FPS & Janks (typically games & media only)
* User events (time-taken for specific user actions like a login or a checkout).
* CPU & Threads

After capturing raw metrics, a lot of different operations (sum, max, top, count and percentiles) are used for pass-fail criteria. 
An interesting (and fast) approach to start adding this pass-fail criteria are based on previous results, like where you can make sure you app is (at least) as good (or bad) as your previous release was.

Thanks for sharing @sregg !

**Comment by pilot34** [comment]
We've recently published a blog post about how we track performance at Booking.com: 
https://medium.com/booking-com-development/measuring-mobile-apps-performance-in-production-726e7e84072f

We've open sourced our performance suites for both platforms. [iOS one](https://github.com/bookingcom/perfsuite-ios) is a bit more extensive. [Android one](https://github.com/bookingcom/perfsuite-android) has only TTI tracking. They don't contain backend logic, only client-side implementation, because we reuse our extensive backend infra for aggregating, displaying metrics and alerting.

We also have performance CI test, it runs UI test going through the main screens 100 times every hour and display median values of all the performance metrics to Grafana with alerting if they jumped too much. I'm going to describe how it is set up in the next article.

**Reply by philipphofmann** [reply]
@pilot34, we at Sentry had the same problem with slow and frozen frames as you described in your blog post and came up with the same approach as you did, if I'm not mistaken, but we now call it frames delay (https://github.com/getsentry/team-mobile/issues/160) 😄. We already shipped it for iOS and soon for Android, but we still need to display the data in our product.

Thanks for making your code open-source. It will improve tracking mobile app performance for every mobile dev.

**Reply by mkj-is** [reply]
Thanks for publishing this. It's nice to see we are converging on a similar solution.

- We previously had issues using Firebase (it expects a fixed frame rate of 60, which is not the case for ProMotion screens and TVs: https://github.com/firebase/firebase-ios-sdk/issues/10220).
- We were not able to get very useful data for rendering performance from Sentry, so I am looking forward to seeing how this "frames delay" improves the tracking.
- The issue with hang rate and hitch ratio is that it changes depending on the user behavior a lot. And we are not able to detect any regression per feature. As mentioned in the blog post…

**Reply by philipphofmann** [reply]
> we are not able to detect any regression per feature.

We plan to build a metric for frames delay per screen so you can track regressions. In the beginning, it will only be available for screen loads. In the next couple of months, we will add [Sentry developer metrics](https://develop.sentry.dev/delightful-developer-metrics/) to the mobile SDKs, and then we can also add a frames delay metric for an entire screen. I added this to our internal roadmap, but can't give an ETA. Thanks for the inspiration, @mkj-is.
